nkms/client.py

```diff
@@ -200,7 +241,13 @@ def open(self, pubkey=None, path=None, mode='r', fd=None, algorithm=None):
 
         If pubkey is not set, we're working on our own files.
         """
-        pass
+        file_path = fd or path
+        try:
+            with open(file_path, mode=mode) as f:
+                enc_data = f.read()
```

Comment:
Yes, exactly, I wanted the file-like interface. The file could potentially be bulky. I think, nacl can decrypt streams, right?

Reply:
NaCl can, I think. You have to pass in the `nonce` param when performing decryption otherwise I think it tries to pull it out of the ciphertext.

Reply:
So, why don't we put nonce at the front (or with the header)?

Reply:
That's what my plan was after I get this interface working and tested.

Reply:
I even think: why not even allow to open the file for writing (when it's for our own pubkey)? In this case, we don't have to keep all the ciphertext in memory

Reply:
I think, we might leave both options here as different methods (file-like interface open() and full decrypted data decrypt()). The full decrypted data will most likely be when the path is not the path to the real file but just, say, "path" to the chat room

Reply:
Why would we need to open a file that we're reading for decryption in a write mode?

Reply:
For write mode: it's if either we create a new one, or we want to add new data to it (as the owner)

Reply:
Ah, okay.

However, if we're adding new data to it, we need to decrypt it and reencrypt it again with a different random nonce.

Reply:
When you're not over-writing but appending data, cannot you use the same nonce for the whole thing in stream ciphers? As if you continued adding blocks to the same file

Reply:
No, it causes the keystream to repeat which can allow someone to break the cipher.

Reply:
I mean, if you add data to an encrypted file - wouldn't it be the same as if you encrypted a larger file to start with (with that nonce)?

Reply:
Yes, but any mistake made when writing the data could allow the cipher to be broken.

You're essentially adding a *huge* attack surface on the schema and that's always a horrible idea. It would be better to reuse the key, generate a new nonce and reencrypt the file again to avoid any major detrimental pit falls.

Reply:
(of course, only if you continue the keystream, not restart it!)

Reply:
Anyway. I guess, let's make it so that you encrypt it all at once at first.

I think ideally we'd like to have it so that each update has its own symmetric key (encrypted by a public key) so that updates to the file will be unreadable by Bob after we revoked the access, even if he saved the symmetric key. And that requires us to not update the old ciphertext.